File: pe/nonvisualstudio/clojure/Implementation/Problem0009.clj
(ns pe.clojure.implementation.problem0009)

(load-file "C:/Users/Thomas/Documents/GitHub/practice/utilities/nonvisualstudio/clojure/Implementation/Sequences.clj")
(use '[utilities.sequences :as util.seq])

(defn euclidsFormula [m n]
  (let [a (- (* m m) (* n n))
        b (* 2 m n)
        c (+ (* m m) (* n n))]
    [a b c]))

(defn findFirstTripleThatSumsTo [target]
  (loop [m 2
         n 1]
    (let [triple (euclidsFormula m n)
          sumOfTriple (util.seq/sum triple)]
      (cond
        (= sumOfTriple target)
          triple
        (= m n)
          (recur (inc m) 1)
        :else
          (recur m (inc n))))))

(defn productOfTheFirstTripleThatSumsTo [target]
  (->>
    (findFirstTripleThatSumsTo target)
    (util.seq/product)))

File: pe/visualstudio/PE.JavaScript/Implementation/Problem0009.js
var javascriptPE;
(function (javascriptPE) {
    (function (implementation) {

        var problem0009 = (function() {

            function problem0009(listHelpers) {
                this.listHelper = listHelpers;
            }

            problem0009.prototype.euclidsFormula = function(m, n) {
                var mSquared = m * m;
                var nSquared = n * n;

                var a = mSquared - nSquared;
                var b = m * n * 2;
                var c = mSquared + nSquared;

                return [a, b, c];
            };

            problem0009.prototype.findFirstTripleThatSumsTo = function(target) {
                var m = 2;
                var n = 1;

                var answer;

                while (answer == undefined) {
                    var euclidsResult = this.euclidsFormula(m, n);

                    if (this.listHelper.sum(euclidsResult) === target) {
                        answer = euclidsResult;
                    };

                    if (m == n) {
                        m++;
                        n = 1;
                    } else {
                        n++;
                    };
                };

                return answer;
            };

            problem0009.prototype.productOfTheFirstTripleThatSumsTo = function(target) {
                var tripleThatSums = this.findFirstTripleThatSumsTo(target);
                return this.listHelper.product(tripleThatSums);
            };
           

            return problem0009;
        })();
        implementation.problem0009 = problem0009;

    })(javascriptPE.implementation || (javascriptPE.implementation = {}));
})(javascriptPE || (javascriptPE = {}));
